src/components/structures/TimelinePanel.js

```diff
@@ -291,6 +299,21 @@ var TimelinePanel = React.createClass({
         this.forceUpdate();
     },
 
+    onLocalEchoUpdated: function(ev, room, oldEventId) {
+        if (this.unmounted) return;
+
+        // ignore events for other rooms
+        if (room !== this.props.room) return;
+
+        // Once the remote echo for an event arrives, we need to turn the
+        // greyed-out event black.  When the localEchoUpdated event is raised,
+        // the nested 'event' property within one of the events in
+        // _timelineWindow will have been replaced with the new event. So
+        // all that is left to do here is to make the message-panel re-render.
+        this.forceUpdate();
```

Comment:
Is it really only the message panel that needs to update here? If so, could the message panel just listen for the event instead?